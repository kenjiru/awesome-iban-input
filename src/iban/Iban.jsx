import React, {Component} from 'react';
import PropTypes from 'prop-types';

const GROUP_SIZE = 4;

class Iban extends Component {
    static formatIbanValue(ibanValue) {
        let formattedStr = '';
        const numGroups = ibanValue.length / GROUP_SIZE;

        for (let i = 0; i < numGroups; i += 1) {
            formattedStr += ibanValue.substring(i * GROUP_SIZE, (i + 1) * GROUP_SIZE);

            if (i < numGroups - 1) {
                formattedStr += ' ';
            }
        }

        return formattedStr;
    }

    static getIbanValue(ibanString) {
        return ibanString.replace(/\s*/g, '');
    }

    constructor(props) {
        super(props);

        this.state = {
            formattedIban: Iban.formatIbanValue(props.value),
        };
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.value !== nextProps.value) {
            this.setState({
                formattedIban: Iban.formatIbanValue(nextProps.value),
            });
        }
    }

    handleChange = (ev) => {
        this.setState({
            formattedIban: ev.target.value,
        });

        const ibanValue = Iban.getIbanValue(ev.target.value);
        this.props.onChange(ibanValue);
    };

    render() {
        return (
            <div className="iban">
                <input onChange={this.handleChange} value={this.state.formattedIban}/>
            </div>
        );
    }
}

Iban.defaultProps = {
    value: '',
};

Iban.propTypes = {
    value: PropTypes.string,
    onChange: PropTypes.func.isRequired,
};

export default Iban;
