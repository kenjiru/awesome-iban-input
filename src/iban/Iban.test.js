import React from 'react';
import shallowToJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Iban from './Iban';

describe('Iban component', () => {
    it('should validate snapshot', () => {
        const changeHandler = jest.fn();

        const output = shallow(<Iban value="AT1231212312" onChange={changeHandler}/>);
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    it('should use the correct transformation methods', () => {
        const ibanValue = 'AT1234567890';
        const ibanStr = 'AT12 3456 7890';

        expect(Iban.formatIbanValue(ibanValue)).toEqual(ibanStr);
        expect(Iban.getIbanValue(ibanStr)).toEqual(ibanValue);
    });

    it('should call the onChange prop when the input changes', () => {
        const initialIbanValue = 'AT1234567890';
        const changedIbanStr = 'AT12 3456 1234';
        const changedIbanValue = 'AT1234561234';

        const changeHandler = jest.fn();

        const output = shallow(<Iban value={initialIbanValue} onChange={changeHandler}/>);

        const input = output.find('input');

        expect(input).toBeDefined();

        input.simulate('change', {
            target: {
                value: changedIbanStr,
            },
        });

        expect(changeHandler).toBeCalledWith(changedIbanValue);
    });

    it('shoudl update the input when we receive a new value', () => {
        const initialIbanValue = 'AT1234567890';
        const changedIbanValue = 'AT1234561234';
        const changedIbanStr = 'AT12 3456 1234';

        const changeHandler = jest.fn();

        const output = shallow(<Iban value={initialIbanValue} onChange={changeHandler}/>);

        output.setProps({
            value: changedIbanValue,
        });

        expect(output.state().formattedIban).toEqual(changedIbanStr);
    });
});
