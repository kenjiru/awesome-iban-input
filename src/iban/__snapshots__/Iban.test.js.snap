// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Iban component validates the snapshot 1`] = `
<div
  className="iban"
>
  <FlashingInput
    isValid={true}
    onChange={[Function]}
    onTab={[Function]}
    value="AT00 0000"
  />
  <span />
</div>
`;
